### DISCUSSION TITLE
The CPU usage of subscribers is too high.

### DISCUSSION THREAD

**Discussion by zhuohuawang** [OpenDDS/OpenDDS#4284] [Q&A]
![dds](https://github.com/OpenDDS/OpenDDS/assets/75567569/c8c5ae56-2bcc-40e8-a4c9-06e6d77ededb)
As shown in the figure. 
When I run a Publisher and a Subscriber application, Publisher cpu usage averages around 5%, but it's not consistent, and on rare occasions it can reach 20%, which is the range I can accept. When running the Subscriber program and receiving data, the average Subscriber cpu usage reaches 15%. This cpu usage I think should be considered an anomaly, considering that I only published 5 topics and only had one subscription per topic.
I want to know whether the Subscriber cpu usage can be reduced by configuring Qos. I tried almost every Qos that I thought had the potential to affect cpu usage, with little success.

**Comment by zhuohuawang** [comment]
Add: Publisher sends data size 15Byte, and sleeps 5ms after each send.

**Reply by jrw972** [reply]
The `durability_service` parts of the writer qos are not used unless the durability kind is `DDS::TRANSIENT_LOCAL_DURABILITY_QOS` on the writer side.

Using history depth and resource limits on the reader leads to additional overhead in the reader.  You should try restoring these to their original values and use `take` instead of `read`.

You should also consider the design of your application.  If you have five threads writing every 5 milliseconds, then why not combine all of the topics into a single topic that is written once every 5 milliseconds?

As another experiment, slow your writers down by waiting 10 milliseconds and see how that affects the CPU of the reader.

**Reply by zhuohuawang** [reply]
I wrote 5 threads because I needed to simulate a real business scenario. I will have more subscribers in the future. So I'm doing a feasibility test right now.
According to your suggestion, I changed the durability kind of writer_qos to DDS::TRANSIENT_LOCAL_DURABILITY_QOS, and used take to get data, but the result didn't improve much. Subscriber cpu usage is still more than 12%.
For this, I wonder why the CPU usage of Subscriber is much higher than that of Publisher. Does Subscriber need any time-consuming operation besides receiving data?

**Reply by jrw972** [reply]
In general, the subscriber has to process the message, store it in a cache, and invoke listeners and waitsets.  The code for receiving a more complicated that writing so I'm not surprised there is a difference.
